truffle/src/main/java/org/jruby/truffle/parser/BodyTranslator.java

```diff
@@ -1826,14 +1826,17 @@ public RubyNode visitHashNode(HashParseNode node) {
         final SourceSection fullSourceSection = sourceSection.toSourceSection(source);
 
         final List<RubyNode> hashConcats = new ArrayList<>();
+        final List<Boolean> hashConcatsIsMerge = new ArrayList<>();
 
         final List<RubyNode> keyValues = new ArrayList<>();
 
         for (KeyValuePair<ParseNode, ParseNode> pair: node.getPairs()) {
             if (pair.getKey() == null) {
                 final RubyNode hashLiteralSoFar = HashLiteralNode.create(context, fullSourceSection, keyValues.toArray(new RubyNode[keyValues.size()]));
                 hashConcats.add(hashLiteralSoFar);
+                hashConcatsIsMerge.add(false);
                 hashConcats.add(HashCastNodeGen.create(context, fullSourceSection, pair.getValue().accept(this)));
+                hashConcatsIsMerge.add(true);
```

Comment:
I'll try refactoring to use a new node. I think it would match the behavior I am looking for.